### DISCUSSION TITLE
Inverted Logic Switches

### DISCUSSION THREAD

**Discussion by WinstonWoof** [WASasquatch/was-node-suite-comfyui#135] [Ideas]
Hi, Would it be possible to have inverted Logic switch options please.

Ie rather than switching between 2 inputs to one output as per existing

![image](https://github.com/WASasquatch/was-node-suite-comfyui/assets/128506111/68241499-8328-493f-8216-e68c2fd0b07d)

to switch 1 input to 2 different outputs (but dont have both mandatory)

This would be a smarter way to implement "feature " switching then for example my current method of using jumpers.

![image](https://github.com/WASasquatch/was-node-suite-comfyui/assets/128506111/6b8656f8-fc41-4fec-8693-78292ab700fe)

**Comment by WASasquatch** [comment]
I still don't think I follow what you mean. How do you get *two* inputs from one without it just being a duplicate?

**Reply by WinstonWoof** [reply]
No 

One Input with 2 Outputs so that for example by using the Boolean switch you could (if it was an Image one) switch an Image output to either a Preview or to a Save step.

Same for Latent & Text 

![image](https://github.com/WASasquatch/was-node-suite-comfyui/assets/128506111/460877e6-f1d8-4c61-adce-9b4a36462aaf)

**Comment by WASasquatch** [comment]
Oooh, I get it now. Sorry, I don't know why that was hard to understand.
It's been a stressful weekend. I'll add that to the to-do.

**Reply by WinstonWoof** [reply]
Pictures speak a thousand words :)

Go pour a drink of your choice ;o)

**Reply by WASasquatch** [reply]
I think there may be an issue with this idea. ComfyUI always expects a valid output for things connected. So if one is valid and other is not sending it would error.

**Reply by WinstonWoof** [reply]
Errrm I already do this with "jumpers" where I can leave them open with no errors

Eg I "break the link to Ksamplers etc like this

![image](https://github.com/WASasquatch/was-node-suite-comfyui/assets/128506111/8fcb58a2-4bf7-467a-bff6-7bf9e6a2221c)

Yes you do get errors in the console but I would call them warning rathe rthan eerrors as the workflow continues normally

![image](https://github.com/WASasquatch/was-node-suite-comfyui/assets/128506111/e6fb80e0-80e5-4bf9-83ab-e43f3e2322b2)


And yes I know I could just "mute" Nodes but that isn't as visually obvious as a Jumper or a switch

**Reply by WASasquatch** [reply]
Yes, because they are simply unhooked. A custom node can't just tell the front end "this is unhooked" when it isn't.

**Reply by WinstonWoof** [reply]
Ok so with that proviso in mind how difficult would it be to invert the output so  it was one in too 2 Out selectable by a boolean switch?

Use cases could be switching image output between a preview & save node or switching a latent feed to alternate ksamplers depending on the workflow

**Reply by WASasquatch** [reply]
You can't without sending actual valid data or it will try to access NoneType as if it were the data.

**Comment by VikingTechLLC** [comment]
If I write it up as so, I do get an image passthrough to image_a with boolean = 1, but yeah, when I try and make boolean = 0 I get the NoneType error.

`class WAS_Image_Input_Switch:
    def __init__(self):
        pass

    @classmethod
    def INPUT_TYPES(cls):
        return {
            "required": {
                "image": ("IMAGE",),
                "boolean_number": ("NUMBER",),
            }
        }

    RETURN_TYPES = ("IMAGE",)
    RETURN_NAMES = ("image_a", "image_b") 
    FUNCTION = "image_input_switch"

    CATEGORY = "WAS Suite/Logic"

    def image_input_switch(self, image, boolean_number=1):

        if int(boolean_number) == 1:
            return (image,)
        else:
            return (None,)`



Traceback (most recent call last):
  File "C:\Users\chris\ComfyUI_windows_portable\ComfyUI\execution.py", line 145, in recursive_execute
    output_data, output_ui = get_output_data(obj, input_data_all)
  File "C:\Users\chris\ComfyUI_windows_portable\ComfyUI\execution.py", line 75, in get_output_data
    return_values = map_node_over_list(obj, input_data_all, obj.FUNCTION, allow_interrupt=True)
  File "C:\Users\chris\ComfyUI_windows_portable\ComfyUI\execution.py", line 68, in map_node_over_list
    results.append(getattr(obj, func)(**slice_dict(input_data_all, i)))
  File "C:\Users\chris\ComfyUI_windows_portable\ComfyUI\nodes.py", line 1141, in save_images
    full_output_folder, filename, counter, subfolder, filename_prefix = folder_paths.get_save_image_path(filename_prefix, self.output_dir, images[0].shape[1], images[0].shape[0])
TypeError: 'NoneType' object is not subscriptable


(not super fluent in python, just seeing what I could do)

**Comment by WASasquatch** [comment]
You need to have two image return types `RETURN_TYPES = ("IMAGE", "IMAGE")`

**Comment by VikingTechLLC** [comment]
I did a bit more work on it, I got it working, but I did have to alter the nodes.py file as well, specifically the class SaveImage:.

Here's the new Image Output Switch:
`class WAS_Image_Output_Switch:
    def __init__(self):
        pass

    @classmethod
    def INPUT_TYPES(cls):
        return {
            "required": {
                "image": ("IMAGE",),
                "boolean_number": ("NUMBER",),
            }
        }

    RETURN_TYPES = ("IMAGE", "IMAGE")
    RETURN_NAMES = ("image_a", "image_b") 
    FUNCTION = "image_output_switch"

    CATEGORY = "WAS Suite/Logic"

    def image_output_switch(self, image, boolean_number=1):
        placeholder_image = np.empty((0, 0)) #placeholder image

        if int(boolean_number) == 1:
            return (image, placeholder_image)
        else:
            if boolean_number == 0:
                return (placeholder_image, image)`
                
                
And here's the edited nodes.py  class SaveImage:
`class SaveImage:
    def __init__(self):
        self.output_dir = folder_paths.get_output_directory()
        self.type = "output"
        self.prefix_append = ""

    @classmethod
    def INPUT_TYPES(s):
        return {"required": 
                    {"images": ("IMAGE", ),
                     "filename_prefix": ("STRING", {"default": "ComfyUI"})},
                "hidden": {"prompt": "PROMPT", "extra_pnginfo": "EXTRA_PNGINFO"},
                }

    RETURN_TYPES = ()
    FUNCTION = "save_images"

    OUTPUT_NODE = True

    CATEGORY = "image"

    def save_images(self, images, filename_prefix="ComfyUI", prompt=None, extra_pnginfo=None):
        filename_prefix += self.prefix_append
        results = []
        for image in images:
            if image.size == 0:
                # Skip saving empty images
                continue

            i = 255. * image.cpu().numpy()
            img = Image.fromarray(np.clip(i, 0, 255).astype(np.uint8))
            metadata = PngInfo()
            if prompt is not None:
                metadata.add_text("prompt", json.dumps(prompt))
            if extra_pnginfo is not None:
                for x in extra_pnginfo:
                    metadata.add_text(x, json.dumps(extra_pnginfo[x]))
                    
            full_output_folder, filename, counter, subfolder, filename_prefix = folder_paths.get_save_image_path(filename_prefix, self.output_dir, images[0].shape[1], images[0].shape[0])

            file = f"{filename}_{counter:05}_.png"
            img.save(os.path.join(full_output_folder, file), pnginfo=metadata, compress_level=4)
            results.append({"filename": file, "subfolder": subfolder, "type": self.type})
            counter += 1

        return {"ui": {"images": results}}
`
![brave_JnGYg3FM1Q](https://github.com/WASasquatch/was-node-suite-comfyui/assets/84103587/f4385679-c980-40f4-b678-ffb4217e6c7c)

**Comment by WASasquatch** [comment]
Yeah but that isn't applicable. The same thing would need to be applied to all nodes. More a thing to fix on the front end to handle any nodes dynamically.

**Reply by VikingTechLLC** [reply]
hmmmm, yeah I now see what you mean.  I did the fix for the "image save" node only....
Tried to work on similar output for a latent output switch, but got that error using an "upscale latent by" node...so I'd have to do that...plus anything else.
